UTRAnnotator.pm

```diff
@@ -0,0 +1,1286 @@
+      }
+
+      $uAUG_lost_evidence= (exists $self->{uORF_evidence})? $self->find_uorf_evidence($UTR_info,$chr,$start_pos): "NA";
+
+      my %uORF_effect = (
+          "uAUG_lost_type" => $uAUG_lost_type,
+          "uAUG_lost_CapDistanceToStart" =>$start_pos,
+          "uAUG_lost_DistanceToCDS" => $uAUG_lost_DistanceToCDS,
+          "uAUG_lost_DistanceToStop" => $uAUG_lost_DistanceToSTOP,
+          "uAUG_lost_KozakContext" => $uAUG_lost_KozakContext,
+          "uAUG_lost_KozakStrength" => $uAUG_lost_KozakStrength,
+          "uAUG_lost_Evidence" => $uAUG_lost_evidence,
+        );
+
+      $output_flag = $output_flag? $output_flag."&"."5_prime_UTR_premature_start_codon_loss_variant": "5_prime_UTR_premature_start_codon_loss_variant";
+      $output_effects = $output_effects? $output_effects."&".$self->transform_hash_to_string(\%uORF_effect):$self->transform_hash_to_string(\%uORF_effect);
+    }
+  }
+
+  $result{'uAUG_lost_flag'} = $output_flag;
+  $result{'uAUG_lost_effect'} = $output_effects;
+
+  return \%result;
+}
+
+sub uFrameshift {
+
+  # Description: annotate if a five_prime_UTR_varint create a frameshift in existing uORFs
+
+  my ($self, $variant_info, $UTR_info, $mut_pos, $end_pos, $mut_utr_seq, $existing_uORF, $mut_uORF, $start) = @_;
+
+  my $chr = $variant_info->{chr};
+  my $pos = $variant_info->{pos};
+  my $ref = $variant_info->{ref};
+  my $alt = $variant_info->{alt};
+
+  my @sequence = split //, $UTR_info->{seq};
+  my $strand = $UTR_info->{strand};
+  my $utr_length = @sequence;
+
+  #return annotators
+  my $uFrameshift_ref_type = ""; # the type of uORF with the reference allele
+  my $uFrameshift_ref_type_length = ""; # the length of the uORF with of the reference allele.
+  my $uFrameshift_StartDistanceToCDS = ""; # the distance between the start codon of the disrupted uORF and CDS
+  my $uFrameshift_alt_type = ""; # the type of uORF with the alternative allele
+  my $uFrameshift_alt_type_length = ""; # the length of the uORF with of the alternative allele
+  my $uFrameshift_KozakContext = ""; # the Kozak context sequence of the disrupted uORF
+  my $uFrameshift_KozakStrength = ""; # the Kozak strength of the disrupted uORF
+  my $uFrameshift_evidence = ""; # whehter there is translation evidence for the disrupted uORF
+
+  #indicate whether the variant ever introduce a frameshift variant
+  my $flag_uORF;
+  my $output_flag = "";
+
+  my $current_kozak="";
+  my $current_kozak_strength="";
+
+  my %result = (); #result is a hash table with two elements: $flag and $output_effects
+  my $output_effects="";
+
+  my $ref_coding = $self->{ref_coding};
+  my $alt_coding = $self->{alt_coding};
+
+  #skip alleles with same length
+  return {} unless(length($ref_coding) ne length($alt_coding));
+
+  #if it's a deletion at the boundary of exon and intron, we would skip the annotation
+
+  my @mut_utr_seq = split //,$mut_utr_seq;
+  my $length = @mut_utr_seq;
+
+  # my @start = @{$self->get_ATG_pos(\@sequence)};
```

Comment:
I forgot to remove this comment, you can remote it:
`# my @start = @{$self->get_ATG_pos(\@sequence)};`